lib/stages/basebandReadout.cpp

```diff
@@ -212,54 +207,69 @@ void basebandReadout::listen_thread(const uint32_t freq_id, basebandReadoutManag
                 }
             }
 
-            dump_to_write = std::make_unique<basebandDumpData>(data);
-            lock.unlock();
-            ready_to_write.notify_one();
+            // we are done copying the samples into the readout buffer
+            mgr.ready({dump_status, data});
         }
     }
 }
 
-void basebandReadout::write_thread(basebandReadoutManager& mgr) {
+void basebandReadout::writeout_thread(basebandReadoutManager& mgr) {
 
     while (!stop_thread) {
-        std::unique_lock<std::mutex> lock(dump_to_write_mtx);
-
-        // This will reset `dump_to_write` to nullptr so that even if this
-        // raises an exception, it will be empty until `listen_thread` copies
-        // the next request's data
-        while (!dump_to_write) {
-            ready_to_write.wait(lock);
-            if (stop_thread)
-                return;
-        }
-
-        auto data = std::move(dump_to_write);
-
         auto next_request = mgr.get_next_ready_request();
-        basebandDumpStatus& dump_status = std::get<0>(next_request);
+        if (!next_request)
+            continue;
+
+        basebandDumpStatus& dump_status = std::get<0>(std::get<0>(*next_request));
+        basebandDumpData dump_data = std::get<1>(std::get<0>(*next_request));
         // Sanity check
-        if (dump_status.request.event_id != data->event_id) {
+        if (dump_status.request.event_id != dump_data.event_id) {
             ERROR("Mismatched event ids: {:d} - {:d}", dump_status.request.event_id,
-                  data->event_id);
+                  dump_data.event_id);
             throw std::runtime_error("Mismatched id - abort");
         }
-        std::mutex& request_mtx = std::get<1>(next_request);
-        readout_in_progress_metric.labels({std::to_string(data->freq_id)}).set(1);
+        std::mutex& request_mtx = std::get<1>(*next_request);
+        readout_in_progress_metric.labels({std::to_string(dump_data.freq_id)}).set(1);
+
+        // first, get read access to the underlying BipBuffer segment
+        BipBufferReader reader(data_buffer);
+        auto rr = reader.access(dump_data.reservation_length);
+        if (!rr) {
+            ERROR("No matching buffer reader reservation for the baseband dump data");
+            throw std::runtime_error("Not able to get reader reservation");
+        }
+
+        // Sanity check that the read reservation points to the same memory used by the writer
+        DEBUG("Read reservation: starting at {}, length {}",
+              rr->data.data() - data_buffer.data.get(), rr->length);
```

Comment:
Does this account for the possible up-to 15 byte offset of the span compared to the reservation?

Reply:
Reservation includes the alignment offset, that's why the sanity check has to align the read reservation before comparing it to the `data` pointer in the `basebandDumpData` structure.